src/Microsoft.PowerShell.Commands.Utility/commands/utility/group-object.cs

```diff
@@ -22,28 +24,30 @@ internal static class PSTuple
         /// <summary>
         /// ArrayToTuple is a helper method used to create a tuple for the supplied input array.
         /// </summary>
+        /// <typeparam name="T">The first generic type parameter.</typeparam>
         /// <param name="inputObjects">Input objects used to create a tuple.</param>
         /// <returns>Tuple object.</returns>
-        internal static object ArrayToTuple(object[] inputObjects)
+        internal static object ArrayToTuple<T>(IList<T> inputObjects)
         {
             Diagnostics.Assert(inputObjects != null, "inputObjects is null");
-            Diagnostics.Assert(inputObjects.Length > 0, "inputObjects is empty");
+            Diagnostics.Assert(inputObjects.Count > 0, "inputObjects is empty");
 
             return ArrayToTuple(inputObjects, 0);
         }
 
         /// <summary>
         /// ArrayToTuple is a helper method used to create a tuple for the supplied input array.
         /// </summary>
-        /// <param name="inputObjects">Input objects used to create a tuple</param>
+        /// <typeparam name="T">The first generic type parameter.</typeparam>
+        /// <param name="inputObjects">Input objects used to create a tuple.</param>
         /// <param name="startIndex">Start index of the array from which the objects have to considered for the tuple creation.</param>
         /// <returns>Tuple object.</returns>
-        internal static object ArrayToTuple(object[] inputObjects, int startIndex)
+        internal static object ArrayToTuple<T>(IList<T> inputObjects, int startIndex)
         {
             Diagnostics.Assert(inputObjects != null, "inputObjects is null");
-            Diagnostics.Assert(inputObjects.Length > 0, "inputObjects is empty");
+            Diagnostics.Assert(inputObjects.Count > 0, "inputObjects is empty");
```

Comment:
This is a fairly common pattern in the engine, that there are asserts, but no backing runtime checks. I have always wondered about it...

Reply:
I guess the thinking is that since it's an internal class, the usage can be demonstrated to be safe and the error checking skipped.

Reply:
Since we don't have an official policy around Assert, I'll not hold up the PR for it and you can ignore the rest. :)

I've been involved in three different threads recently where Dbg.Assert in internal code is hit in debug builds.  In two cases, crashes were occurring the release build and the third the lack of the check was masking a bug in another component.

IMHO: Since we don't test debug builds, Dbg.Assert is a placebo at best without a release check.